CRM/Core/Invoke.php

```diff
@@ -324,7 +324,12 @@ public static function runItem($item) {
             $object = new CRM_Import_Controller($title, $pageArgs ?? []);
           }
           else {
-            $object = new $item['page_callback']($title, TRUE, $mode, NULL, $addSequence);
+            try {
```

Comment:
@eileenmcnaughton I think if we were to go down this path we should be further up about here https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Invoke.php#L38 to run before the Catch (Exception)

Reply:
ok - I was hoping to discuss with Tim in a bit - but not wanting to divert his attention to the rc until he has dealt with the release (hence no ping)

Reply:
I do think it's useful to have some mechanism for translating exceptions to responses.

FWIW, you should be able to do that without patching  `CRM_Core_Invoke` -- by using `hook_civicrm_unhandled_exception` . The upshot is that anyone listening for events (extensions and/or CMS-integrations) can choose how to render `UnavailableFormException` (or `PermissionDeniedException` or `NotFoundException` or whathaveyou).

Just to spitball/demonstrate, one can do a listener like:

```php
/**
 * Show unhandled exceptions with eponymous Smarty templates.
 *
 * To define `Civi\Foo\MyException`:
 *    - Create `Civi/Foo/MyException.php` and implement `getStatusCode(): int`.
 *    - Create `templates/Civi/Foo/MyException.tpl`
 *
 * @service civi.exceptions.smarty
 */
class SmartyExceptionRenderer extends AutoService implements EventSubscriberInterface {
  public function getSubscribedEvents() {
    return ['hook_civicrm_unhandled_exception' => ['&onException', -1000]];
  }
  public function onException(\Throwable $e) {
    $statusCode = is_callable([$e, 'getStatusCode']) ? $e->getStatusCode() : 500;
    $className = get_class($e);
    $tplName = preg_replace(';[_\\];', DIRECTORY_SEPARATOR', $className) . '.tpl';
    $pageContent = CRM_Core_Smarty::singleton()->fetchWith($tplName, [
      'exception' => $e,
    ]);
    $html = CRM_Utils_System::theme($pageContent);

    CRM_Utils_System::sendResponse(new Response($statusCode.... 'text/html'.... $html));

    // I'm not sure if CRM_Utils_System::theme() works consistently. If not, then
    // we may need to do some juggling in CRM_Core_Invoke.
  }
}
```

But that's getting ahead of ourselves. For an exception-based approach, the first question is taxonomy of exceptions. (Eileen mentioned have some mixed experience with this?) If the taxonomy is open/big, then I think a generic mapping (like above) is fine. But if it's closed/small, then something simpler may do.